### ISSUE TITLE
Query: Cross join with table having TPH mapping causes discriminator to be ignored producing incorrect results

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#8375]
Comparing_entities_using_Equals_inheritance

```cs
                var query = from g in ctx.Gears
                            from o in ctx.Gears.OfType<Officer>()
                            where g.Equals(o)
                            orderby g.Nickname
                            select new { Nickname1 = g.Nickname, Nickname2 = o.Nickname };
```

produces:

```sql
SELECT [g].[Nickname] AS [Nickname1], [o].[Nickname] AS [Nickname2]
FROM [Gear] AS [g]
CROSS JOIN [Gear] AS [o]
WHERE [g].[Discriminator] IN (N'Officer', N'Gear') AND (([g].[Nickname] = [o].[Nickname]) AND ([g].[SquadId] = [o].[SquadId]))
ORDER BY [Nickname1]
```

discriminator predicate for Officer in the second collection is being ignored so we return too many results

**Comment by smitpatel** [comment]
#8094 would fix this too.

**Comment by ajcvickers** [comment]
Blocked by #8094

**Comment by smitpatel** [comment]
Fixed in #17724
